OSSM-4818 Set dynamically-determined UID/GID in gateways

yannuil opened the pull request:
Pick changes only related to the gateway deployment controller from the commit below.

* Remove the need of anyuid permission on OpenShift (#45394)

- Remove the hard-coded usages of `runAsUser` and `runAsGroup` in charts
- Where necessary, replace them with a helm field `.ProxyUID` and `.ProxyGID`.
- When running in OpenShift, these fields will be set at runtime based on annotations present on the pod's namespace.
- When not on OpenShift (strictly speaking, when such annotations are not present), these values fallback to the current, defauls value of 1337.

With this change we can remove the extra steps related to `anyuid` in the [OpenShift setup page](https://istio.io/v1.17/docs/setup/platform-setup/openshift/).

**Please provide a description of this PR:**



**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

yannuil commented:
/test maistra-istio-unit-2-5

yannuil commented:
/test maistra-istio-unit-2-5

yannuil commented:
/test maistra-istio-unit-2-5

yannuil commented:
/test maistra-istio-unit-2-5

yannuil commented:
/test integ-servicemesh-2-5

yannuil commented:
/test integ-servicemesh-2-5

jewertow reviewed (commented):
```
2023-10-24T08:25:25.320798Z	info	gateway	reconciling	gateway=istio-ingress/gateway
2023-10-24T08:25:25.333852Z	error	controllers	error handling istio-ingress/gateway, retrying (retry count: 1): failed to render template: template: inject:39:42: executing "inject" at <.Release.Namespace>: can't evaluate field Release in type gateway.derivedInput	controller=gateway deployment
2023-10-24T08:25:25.333898Z	info	gateway	reconciling	gateway=istio-ingress/gateway
2023-10-24T08:25:25.335147Z	error	controllers	error handling istio-ingress/gateway, retrying (retry count: 2): failed to render template: template: inject:39:42: executing "inject" at <.Release.Namespace>: can't evaluate field Release in type gateway.derivedInput	controller=gateway deployment
2023-10-24T08:25:25.339336Z	info	gateway	reconciling	gateway=istio-ingress/gateway
2023-10-24T08:25:25.341012Z	error	controllers	error handling istio-ingress/gateway, retrying (retry count: 3): failed to render template: template: inject:39:42: executing "inject" at <.Release.Namespace>: can't evaluate field Release in type gateway.derivedInput	controller=gateway deployment
2023-10-24T08:25:25.361347Z	info	gateway	reconciling	gateway=istio-ingress/gateway
2023-10-24T08:25:25.363127Z	error	controllers	error handling istio-ingress/gateway, retrying (retry count: 4): failed to render template: template: inject:39:42: executing "inject" at <.Release.Namespace>: can't evaluate field Release in type gateway.derivedInput	controller=gateway deployment
2023-10-24T08:25:25.403519Z	info	gateway	reconciling	gateway=istio-ingress/gateway
2023-10-24T08:25:25.404985Z	error	controllers	error handling istio-ingress/gateway, and retry budget exceeded: failed to render template: template: inject:39:42: executing "inject" at <.Release.Namespace>: can't evaluate field Release in type gateway.derivedInput	controller=gateway deployment
```

jewertow commented on pilot/pkg/config/kube/gateway/deploymentcontroller.go:
```diff
@@ -298,6 +298,12 @@ func (d *DeploymentController) configureIstioGateway(log *istiolog.Scope, gw gat
+	if d.namespaces != nil {
+		ns = d.namespaces.Get(gw.Namespace, "")
+	}
+	proxyUID, proxyGID := GetProxyIDs(ns)
```
It will not work in multi-tenant mode, because when there is no namespace informer and `ns` is `nil`, `GetProxyIDs` will return 1337, which will cause failure on pod creation, because of missing `anyuid` SCC.

jewertow replied:
So I think that the only option we have is to remove UID/GID 1337 from the kube-gateway template if multi-tenant mode is enabled. Then UID/GID will be set automatically to the lowest value from the assigned range.

jewertow commented on pilot/pkg/config/kube/gateway/deploymentcontroller.go:
```diff
@@ -500,6 +508,8 @@ type TemplateInput struct {
 	KubeVersion122 bool
 	Revision       string
+	ProxyUID       int64
+	ProxyGID       int64
```
These values are not used in kube-gateway.yaml.

yannuil committed:
Set UID/GID in gateway deployments

Partially pick changes from the upstream commits below.

* Remove the need of anyuid permission on OpenShift (#45394)

- Remove the hard-coded usages of `runAsUser` and `runAsGroup` in charts
- Where necessary, replace them with a helm field `.ProxyUID` and
  `.ProxyGID`.
- When running in OpenShift, these fields will be set at runtime based
  on annotations present on the pod's namespace.
- When not on OpenShift (strictly speaking, when such annotations are
  not present), these values fallback to the current, defauls
  value of 1337.

With this change we can remove the extra steps related to `anyuid` in the [OpenShift setup
page](https://istio.io/v1.17/docs/setup/platform-setup/openshift/).

* Fix proxy UID/GID in Gateway templates (#46898)

Co-authored-by: Yann Liu <yannliu@redhat.com>
Co-authored-by: Jonh Wendell <jonh.wendell@redhat.com>
Signed-off-by: Yann Liu <yannliu@redhat.com>

yannuil commented:
/retest

jewertow commented:
You can close this PR for now, because kube-gateway template does not set UID/GID=1337 by default, so we don't need this change.
